lib/attache_rails/model.rb

```diff
@@ -1,150 +1,52 @@
-      end
-
-      def attache_auth_options
-        if ENV['ATTACHE_SECRET_KEY']
-          uuid = SecureRandom.uuid
-          expiration = (Time.now + ATTACHE_UPLOAD_DURATION).to_i
-          hmac = OpenSSL::HMAC.hexdigest(OpenSSL::Digest.new('sha1'), ENV['ATTACHE_SECRET_KEY'], "#{uuid}#{expiration}")
-          { uuid: uuid, expiration: expiration, hmac: hmac }
-        else
-          {}
-        end
-      end
-
-      def attache_options(geometry, current_value, options)
-        {
-          multiple: options[:multiple],
-          class: 'enable-attache',
-          data: {
-            geometry: geometry,
-            value: [*current_value],
-            placeholder: [*options[:placeholder]],
-            uploadurl: ATTACHE_UPLOAD_URL,
-            downloadurl: ATTACHE_DOWNLOAD_URL,
-          }.merge(options[:data] || {}).merge(attache_auth_options),
-        }
-      end
-    end
-  end
   module Model
+    include Attache::API::Model
+
     def self.included(base)
+      # has_one_attache, has_many_attaches
       base.extend ClassMethods
+
+      # `discard` management
       base.class_eval do
         attr_accessor :attaches_discarded
-        after_commit  :attaches_discard!, if: :attaches_discarded
+        after_commit if: :attaches_discarded do |instance|
+          instance.attaches_discard!(instance.attaches_discarded)
+        end
       end
     end
 
-    def attaches_discard!(files = attaches_discarded)
-      files.reject!(&:blank?)
-      files.uniq!
-      if files.present?
-        logger.info "DELETE #{files.inspect}"
-        HTTPClient.post_content(
-          URI.parse(ATTACHE_DELETE_URL),
-          Utils.attache_auth_options.merge(paths: files.join("\n"))
-        )
+    module ClassMethods
+      def attache_setup_column(name)
+        case coltype = column_for_attribute(name).type
+        when :text, :string, :binary
+          serialize name, JSON
+        end
+      rescue Exception
       end
-    rescue Exception
-      raise if ENV['ATTACHE_DISCARD_FAILURE_RAISE_ERROR']
-      logger.warn [$!, $@]
-    end
 
-    module ClassMethods
       def has_one_attache(name)
-        serialize name, JSON
-        define_method "#{name}_options",    -> (geometry, options = {}) { Utils.attache_options(geometry, Utils.array(self.send("#{name}_attributes", geometry)), multiple: false, **options) }
-        define_method "#{name}_url",        -> (geometry) {               self.send("#{name}_attributes", geometry).try(:[], 'url') }
-        define_method "#{name}_attributes", -> (geometry) {               str = self.send(name); Utils.attache_url_for(str, geometry) if str; }
-        define_method "#{name}=",           -> (value)    {
-          new_value = (value.respond_to?(:read) ? Utils.attache_upload_and_get_json(value) : value)
-          okay = JSON.parse(new_value)['path'] rescue nil
-          super(Utils.array(okay ? new_value : nil).first)
-        }
-        define_method "#{name}_discard_was",-> do
```

Comment:
yes. we're doing "mark and sweep"
- during `after_update` we collect obsolete values from `old_values - new_values`
- during `after_destroy` we collect all current values as obsolete

then finally, everything is "swept" at `after_commit`